src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -0,0 +1,394 @@
+  IL_001d:  ldloc.3
+  IL_001e:  ldc.i4.1
+  IL_001f:  ldc.i4.0
+  IL_0020:  newobj     ""System.Index..ctor(int, bool)""
+  IL_0025:  ldc.i4.1
+  IL_0026:  ldc.i4.1
+  IL_0027:  newobj     ""System.Index..ctor(int, bool)""
+  IL_002c:  newobj     ""System.Range..ctor(System.Index, System.Index)""
+  IL_0031:  call       ""char[] System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray<char>(char[], System.Range)""
+  IL_0036:  stloc.1
+  IL_0037:  ldloc.3
+  IL_0038:  ldloc.s    V_4
+  IL_003a:  ldc.i4.1
+  IL_003b:  sub
+  IL_003c:  ldelem.u2
+  IL_003d:  stloc.2
+  IL_003e:  br.s       IL_0042
+  IL_0040:  ldc.i4.1
+  IL_0041:  ret
+  IL_0042:  ldloc.0
+  IL_0043:  ldloc.2
+  IL_0044:  bne.un.s   IL_004d
+  IL_0046:  ldloc.1
+  IL_0047:  call       ""bool X.Test(char[])""
+  IL_004c:  ret
+  IL_004d:  ldc.i4.0
+  IL_004e:  ret
+}
+");
+            verifier.VerifyIL("X.Test(string)", @"
+{
+  // Code size       73 (0x49)
+  .maxstack  4
+  .locals init (char V_0, //first
+                string V_1, //others
+                char V_2, //last
+                string V_3,
+                int V_4)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.3
+  IL_0002:  ldloc.3
+  IL_0003:  brfalse.s  IL_0047
+  IL_0005:  ldloc.3
+  IL_0006:  callvirt   ""int string.Length.get""
+  IL_000b:  stloc.s    V_4
+  IL_000d:  ldloc.s    V_4
+  IL_000f:  ldc.i4.2
+  IL_0010:  bge.s      IL_0019
+  IL_0012:  ldloc.s    V_4
+  IL_0014:  ldc.i4.1
+  IL_0015:  ble.un.s   IL_003a
+  IL_0017:  br.s       IL_0047
+  IL_0019:  ldloc.3
+  IL_001a:  ldc.i4.0
+  IL_001b:  callvirt   ""char string.this[int].get""
+  IL_0020:  stloc.0
+  IL_0021:  ldloc.3
+  IL_0022:  ldc.i4.1
+  IL_0023:  ldloc.s    V_4
+  IL_0025:  ldc.i4.2
+  IL_0026:  sub
+  IL_0027:  callvirt   ""string string.Substring(int, int)""
+  IL_002c:  stloc.1
+  IL_002d:  ldloc.3
+  IL_002e:  ldloc.s    V_4
+  IL_0030:  ldc.i4.1
+  IL_0031:  sub
+  IL_0032:  callvirt   ""char string.this[int].get""
+  IL_0037:  stloc.2
+  IL_0038:  br.s       IL_003c
+  IL_003a:  ldc.i4.1
+  IL_003b:  ret
+  IL_003c:  ldloc.0
+  IL_003d:  ldloc.2
+  IL_003e:  bne.un.s   IL_0047
+  IL_0040:  ldloc.1
+  IL_0041:  call       ""bool X.Test(string)""
+  IL_0046:  ret
+  IL_0047:  ldc.i4.0
+  IL_0048:  ret
+}
+");
+        }
+
+        [Theory]
+        // Index
+        [InlineData("[Index i]")]
+        [InlineData("[Index i, int ignored = 0]")]
+        [InlineData("[Index i, params int[] ignored]")]
+        [InlineData("[params Index[] i]")]
+        [InlineData("[int i]")]
+        public void ListPattern_IndexIndexerPattern(string indexer)
+        {
+            var source = @"
+using System;
+class X
+{
+    public int this" + indexer + @"
+    {
+        get
+        {
+            i.ToString(); // verify argument is usable
+            Console.Write(""this[] "");
+            return 1;
+        }
+    }
+    public int Length => 1;
+    public static void Main()
+    {
+        Console.Write(new X() is { 1 });
+    } 
```

Comment:
Please add a test passing an index (ie. showing that the type is "indexable" in the Index/Ranges sense).
It would also be good to keep some of those error cases instead of removing them entirely.